File: src/tools/oidc.ts
import { type ToolDefinition, type ToolResult } from "../core/types.js";

export function getTools(): ToolDefinition[] {
  const make = (n: string, d: string, s: Record<string, unknown>, fn: (p: any, ctx: any) => Promise<ToolResult>): ToolDefinition => ({
    name: n, description: d, inputSchema: s as any, execute: fn,
  });
  return [
    make("oidc_providers", "List providers. Read-only.", { type:"object", properties:{} },
      async (_p:any, ctx:any) => ({ isSuccess:true, data: await ctx.query("query { oidc { providers { id name enabled } } }") })),
    make("oidc_get", "Get provider details.", { type:"object", properties:{ provider_id:{type:"string"} }, required:["provider_id"] },
      async (p:any, ctx:any) => ({ isSuccess:true, data: await ctx.query("query OidcProvider($id: ID!) { oidc { provider(id: $id) { id name enabled clientId } } }", p) })),
    make("oidc_create", "Create provider.", { type:"object", properties:{ name:{type:"string"}, issuer:{type:"string"}, clientId:{type:"string"}, clientSecret:{type:"string"} }, required:["name","issuer","clientId"] },
      async (p:any, ctx:any) => ({ isSuccess:true, data: await ctx.query("mutation CreateOidc($name: String!, $issuer: String!, $clientId: String!, $clientSecret: String) { oidc { create(name: $name, issuer: $issuer, clientId: $clientId, clientSecret: $clientSecret) { id } } }", p) })),
    make("oidc_update", "Update provider.", { type:"object", properties:{ provider_id:{type:"string"}, name:{type:"string"}, issuer:{type:"string"}, clientId:{type:"string"}, enabled:{type:"boolean"} }, required:["provider_id"] },
      async (p:any, ctx:any) => ({ isSuccess:true, data: await ctx.query("mutation UpdateOidc($id: ID!, $name: String, $issuer: String, $clientId: String, $enabled: Boolean) { oidc { update(id: $id, name: $name, issuer: $issuer, clientId: $clientId, enabled: $enabled) { id } } }", p) })),
    make("oidc_delete", "Delete provider.", { type:"object", properties:{ provider_id:{type:"string"} }, required:["provider_id"] },
      async (p:any, ctx:any) => ({ isSuccess:true, data: await ctx.query("mutation DeleteOidc($id: ID!) { oidc { delete(id: $id) { status } } }", p) })),
  ];
}

File: go/internal/tools/oidc.go
package tools

import (
	"context"
	"fmt"

	"github.com/hektyc/unraid-mcp-server/go/internal/config"
	"github.com/hektyc/unraid-mcp-server/go/internal/guards"
	"github.com/hektyc/unraid-mcp-server/go/internal/mcp"
)

func RegisterTools(s *mcp.Server, cfg *config.Config) {
	s.RegisterTool(&mcp.ToolDef{
		Name:        "oidc_providers",
		Description: "List providers. Read-only.",
		Query:       `query { oidc { providers { id name enabled } } }`,
	})
	s.RegisterTool(&mcp.ToolDef{
		Name:        "oidc_create",
		Description: "Create provider.",
		Query:       `mutation CreateOidc($name: String!, $issuer: String!, $clientId: String!) { oidc { create(name: $name, issuer: $issuer, clientId: $clientId) { id } } }`,
		Params: map[string]string{
			"name": "string", // required=true
			"issuer": "string", // required=true
			"clientId": "string", // required=true
		},
	})
	s.RegisterTool(&mcp.ToolDef{
		Name:        "oidc_delete",
		Description: "Delete provider.",
		Query:       `mutation DeleteOidc($id: ID!) { oidc { delete(id: $id) { status } } }`,
		Params: map[string]string{
			"provider_id": "string", // required=true
		},
	})
}